### PR TITLE
DRA: fix scheduler/resource claim controller race

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

There was a race caused by having to update claim finalizer and status in two different operations:
- Resource claim controller removes allocation, does not yet get to remove the finalizer.
- Scheduler prepares an allocation, without adding the finalizer because it's there.
- Controller removes finalizer.
- Scheduler adds allocation.

This is an invalid state. Automatic checking found this during the execution of the "with translated parameters on single node.*supports sharing a claim sequentially" E2E test, but only when run stand-alone. When running in parallel (as in the CI), the bad outcome of the race did not occur.

#### Special notes for your reviewer:

The fix is to check that the finalizer is still set when adding the allocation. This can be done with a complicated JSON patch (see first commit, but only if you are really, really curious!), but a local retry loop with Update calls is simpler.

The resource claim controller doesn't need this, it can do a normal update which implicitly checks ResourceVersion.

#### Does this PR introduce a user-facing change?
```release-note
DRA: using structured parameters with a claim that gets reused between pods may have led to a claim with an invalid state (allocated without a finalizer) which then caused scheduling of pods using the claim to stop.
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 1]
```diff
@@ -1632,20 +1635,51 @@ func (pl *dynamicResources) bindClaim(ctx context.Context, state *stateData, ind
...
+	// JSON patch does not support that, so we check for the stronger "array is equal".
+	if len(claim.Finalizers) > 0 {
+		parts = append(parts,
+			fmt.Sprintf(`{ "op": "test", "path": "/metadata/finalizers", "value": %q }`, claim.Finalizers),
+		)
+	}
+	// JSON patch can only append to a non-empty array. An empty reservedFor gets
+	// omitted and even if it didn't, it would be null and not an empty array.
+	// Therefore we have to test and add if it's currently empty.
+	reservedForEntry := fmt.Sprintf(`{"resource": "pods", "name": %q, "uid": %q}`, pod.Name, pod.UID)
```
Comment (pohly) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:1657]: While it was fun playing around with JSON patch, I think this is taking it too far...

A simpler, more obvious approach would be to add a retry loop which uses normal Update calls and gets the latest claim on a conflict.

Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go]: Implemented, ready for review again.

**Comment by pohly** [comment]
/assign @kerthcet 

Can you help review?

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 2]
```diff
@@ -1592,72 +1593,89 @@ func (pl *dynamicResources) PreBind(ctx context.Context, cs *framework.CycleStat
...
-			if _, err := pl.clientset.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{}); err != nil {
-				return nil, fmt.Errorf("add finalizer: %v", err)
+		// Do we need to store an allocation result from Reserve?
+		if allocation != nil {
+			if claim.Status.Allocation != nil {
+				return nil, fmt.Errorf("claim %s got allocated elsewhere in the meantime", klog.KObj(claim))
+			}
+
+			// The finalizer needs to be added in a normal update.
+			// If we were interrupted in the past, it might already be set and we simply continue.
```
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:1630]: Should we get for the laster claim to check whether the finalizer has been removed?

Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:1630]: If it has been removed and the `claim` instance is stale, then the `UpdateStatus` below will fail and we retry with a fresh copy of the claim.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 3]
```diff
@@ -1592,72 +1593,89 @@ func (pl *dynamicResources) PreBind(ctx context.Context, cs *framework.CycleStat
...
+			if !slices.Contains(claim.Finalizers, resourcev1alpha2.Finalizer) {
+				claim.Finalizers = append(claim.Finalizers, resourcev1alpha2.Finalizer)
+				updatedClaim, err := pl.clientset.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
+				if err != nil {
+					if apierrors.IsConflict(err) {
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go]: Pushed.

I verified again that the retry loop works: when running `_output/bin/ginkgo -v --focus="with translated parameters on single node.*supports sharing a claim sequentially" ./test/e2e`, the scheduler had to retry once as indicated by the log message when it gets the newer claim.

**Comment by kerthcet** [comment]
/lgtm
/approve

The two commits seems exclusive, will you squash them?
/hold

**Comment by pohly** [comment]
I'd prefer to keep the two commits: the first one can serve as reference for how this would have looked like with JSON patching.

/hold cancel

**Pull request merged**